### DISCUSSION TITLE
Slow operation

### DISCUSSION THREAD

**Discussion by sheepdragon** [argoproj/argo-workflows#6631] [Q&A]
![image](https://user-images.githubusercontent.com/4659052/131518909-2aba956d-10e7-45ec-8ad2-5cd08a065046.png)
Why is such a simple example so time-consuming

**Comment by jrbe228** [comment]
This minimum ~10s delay may be the biggest block from our organization adopting ArgoWF at larger scale. There may be no easy solution given the delays fundamental to Kubernetes itself. It's frustrating when compared to `docker run hello-world` taking milliseconds. 

I tried the WASM plugin as a potential workaround, but got even slower results - https://github.com/Shark/wasm-workflows-plugin/issues/16, at least initially. 

I see other discussions suggesting the [argo-dataflow](https://github.com/argoproj-labs/old-argo-dataflow) (now [numaflow](https://github.com/numaproj/numaflow/tree/main)) project for data processing workflows. Is this the only known alternative? How would we adapt Argo workflows with legacy containers to a data-centric framework?

**Reply by jswxstw** [reply]
> I tried the WASM plugin as a potential workaround, but got even slower results - https://github.com/Shark/wasm-workflows-plugin/issues/16, at least initially.

I also tried HTTP/Plugin as a workaround, and it has indeed significantly reduced the execution time of the workflow.
@jrbe228 Have you reduced these two [controller envs](https://argo-workflows.readthedocs.io/en/latest/environment-variables/#controller): `ARGO_AGENT_PATCH_RATE` and `DEFAULT_REQUEUE_TIME`?

**Reply by jrbe228** [reply]
I don't see any HTTP Plugin in the [directory](https://argo-workflows.readthedocs.io/en/latest/plugin-directory/) but maybe you are using HTTP [Templates](https://argo-workflows.readthedocs.io/en/latest/http-template/)? Also thanks for the reference on controller Env Vars. I can test those soon.

**Reply by jswxstw** [reply]
I initially tried the HTTP Template, but its functionality was quite limited, so later I implemented a custom CLI-based plugin that serves HTTP requests containing CLI parameters.

**Reply by jrbe228** [reply]
Interesting. Reducing the Env Vars ARGO_AGENT_PATCH_RATE and DEFAULT_REQUEUE_TIME did improve performance. They are set to 10s by default. Reducing to 1s caused approx 30% improvement (10s to 7s) in completion time for trivial workflows like [hello-world](https://raw.githubusercontent.com/argoproj/argo-workflows/master/examples/hello-world.yaml). There was no further benefit from reducing the vars into the millisecond range. 

I also tried the [suggested](https://argo-workflows.readthedocs.io/en/latest/http-template/) HTTP Template workflow. The core workflow step "get-google-homepage" completes in 3 to 5 seconds. For some reason the parent workflow continues to "spin" for several minutes , as if still executing. 

Lastly, I retried WASM plugin execution with the new Env Var config. Execution time dropped from ~30 seconds I described in this [issue](https://github.com/Shark/wasm-workflows-plugin/issues/16) to ~5 seconds. A very encouraging result! 

![image](https://github.com/user-attachments/assets/30ff907b-7a81-4132-92d6-5a4090fbb8b2)

**Reply by jrbe228** [reply]
@jswxstw - for your custom CLI-based HTTP plugin, how fast do minimal workflows execute? The best results I see from the WASM plugin around 4 seconds: 
![image](https://github.com/user-attachments/assets/d8bfcae7-8d02-4ec3-8b4e-420d4e1b2d9a)

**Reply by jswxstw** [reply]
Argo is not inherently a low-latency task scheduling system, what you've measured now is already almost the minimum time consumption (the agent pod creation itself takes several seconds).
Also, reducing DEFAULT_REQUEUE_TIME may increase the load of APIServer, as the scale of the workflow increases, the time taken for pod creation and various reconciliation will significantly increase, which is uncontrollable.

**Reply by jrbe228** [reply]
Some latency appears fundamental to Kubernetes. Testing minimal pods takes around 4 seconds on AKS - 
`kubectl run hello-world --image=hello-world --restart=Never --rm -it`

I imagine a bare-metal K8s distro could improve somewhat on latency. One less layer of virtualization. 

Otherwise I would need some kind of server/client framework using WASM pods. I tried to describe it [here](https://github.com/Shark/wasm-workflows-plugin/issues/16#issuecomment-2452092002).
